src/group_theory/order_of_element.lean

```diff
@@ -120,23 +120,189 @@ begin
   { exact ⟨i + 1, dec_trivial, inv_eq_one.1 eq⟩ }
 end
 
-include dec
+end order_of
+
+section order_of
+
+section monoid
+open_locale classical
+variables {α} [monoid α] [decidable_eq α]
 
-/-- `order_of a` is the order of the element `a`, i.e. the `n ≥ 1`, s.t. `a ^ n = 1` -/
-def order_of (a : α) : ℕ := nat.find (exists_pow_eq_one a)
+
+/-- `order_of a` is the order of the element `a`, i.e. the `n ≥ 1`, s.t. `a ^ n = 1` if it exists.
+Otherwise, i.e. if `a` is of infinite order, then `order_of a` is `0` by convention.-/
+noncomputable def order_of (a : α) : ℕ :=
+if h : ∃ n, 0 < n ∧ a ^ n = 1 then nat.find h else 0
```

Comment:
Are you sure you want classical logic here? I think it might be cleaner to leave that up to the user of this definition:
```suggestion
def order_of (a : α) [decidable_pred (λ n, 0 < n ∧ a ^ n = 1)] [decidable (∃ n, 0 < n ∧ a ^ n = 1)] : ℕ :=
if h : ∃ n, 0 < n ∧ a ^ n = 1 then nat.find h else 0

instance order_of.fintype_decidable {α : Type u_1} [group α] [fintype α] (a : α) :
∃ (i : ℕ) (H : i > 0), a ^ i = 1 := decidable.of_true (exists_pow_eq_one α)
```

This means that with the original `fintype` and `dexidable_eq` conditions, this remains computable; but any user can always just `open_locale classical` and everything ought to work ok still.

But perhaps this is unreasonably constructive.

Reply:
I think it makes sense to be computable here. It's nice if Lean can just compute the order of an element in a group with < 10 elements. Such things might actually be useful when proving abstract stuff. (Eg computations in `zmod 4` or something like that.)

Reply:
Thanks for the suggestion. I have no idea at all what the best way is to do this but I feel like there will be users who want to talk about the order of an element in a possibly infinite group. For me, classical logic was the easiest to make this file compile, there might well be better solutions. I'll try your suggestion.

Reply:
I would substitute `decidable_pred (λ n, 0 < n ∧ a ^ n = 1)` for `decidable_eq α`. I do have some concerns that the decidability assumptions could cause some diamond problems, and it might be best to remain classical.

Reply:
@ChrisHughes24, my understanding is that your suggestion makes the diamond problems worse not better.

Reply:
The suggestion of `decidable_eq` or the suggestion of going classical?

Reply:
The suggestion of `decidable_eq`. Obviously going `classical` here will avoid diamond problems elsewhere, but it also means `dec_trivial` can no longer find the order of an element.